Revert changes to type of Content

tectiv3 opened the pull request:
Had to revert this cause this change broke context ("memory" of the model). 
Haven't dug too deep into the cause but simply reverting it back to value instead of reference fixed things.

Calling function is fun though. I was going to add it myself but you were faster ;)

meinside commented:
Sorry for the breaking change, but it was intentional.

The `content` value became optional due to the addition of function calls,
so the best way to distinguish between empty string and null value (or property not returned at all)
was making it a pointer and tag as `omitempty`.

I don't get what you're saying about the 'context' or 'memory of the model',
but I think it might be fixed by nil-checking the pointer first, and
using the 'value' (not the pointer itself) if it is not nil.
If you want, I can help you out with your problem.

tectiv3 commented:
```go
	chat := s.getChat(c.Chat().ID)
	msg := openai.NewChatUserMessage(message)
	system := openai.NewChatSystemMessage(chat.MasterPrompt)

	chat.History = append(chat.History, ChatMessage{Role: msg.Role, Content: msg.Content, ChatID: chat.ChatID})
	history := []openai.ChatMessage{system}
	for _, h := range chat.History {
		history = append(history, openai.ChatMessage{Role: h.Role, Content: h.Content})
	}
...
        s.ai.CreateChatCompletion(chat.ModelName, history, options)
```

In this snippet I am loading messages history from the DB to pass on in a request as a context.
But with the new change of Content being a reference and probably in 
https://github.com/meinside/openai-go/blob/26a5357fe0453aa7adf9b6f234038e93ec810ec6/chat.go#L246

the history is being sent empty. I have no idea why.


Also I don't see why you needed this change. All my code with or without functions works perfectly fine with Content being value and not a reference (this PR).


The only thing not working, unrelated to this change, is that I couldn't get function calls and streaming to work yet.

meinside commented:
Can I see the definition of your `ChatMessage`?

tectiv3 commented:
```go
type ChatMessage struct {
	ID        uint `gorm:"primarykey"`
	CreatedAt time.Time
	UpdatedAt time.Time
	ChatID    int64
	Role      openai.ChatMessageRole `json:"role"`
	Content   string                 `json:"content,omitempty"`
}
```

and before it was simply
```go
type ChatMessage struct {
	ID        uint `gorm:"primarykey"`
	CreatedAt time.Time
	UpdatedAt time.Time
	ChatID    int64
	openai.ChatMessage
}

meinside commented:
As you know, the type of `Content` was changed from `string` to `*string`, so in this case you should use like:

```
ChatMessage{Role: msg.Role, Content: *msg.Content, ChatID: chat.ChatID}
```

for passing the value, not the pointer itself.

You can also modify your type to pointer, but that depends on you.

If you don't get why its type was changed to *string, please search for some articles about JSON unmarshalling in golang.

It was just clarifying things returned & parsed from API responses.

tectiv3 commented:
I’m using my fork:) from this PR.
My point is there is no need in making Content a reference.

meinside commented:
Yeah, I know your point and I don't want to make breaking changes :-)
But I wanted to follow up the changes on the OpenAI's side,
and make it clear - if it was intentionally an empty string, or if the property was not returned at all from API.
Optional value types in structs (in this case, string) need to be pointers for that, and that's the one of the best practices in golang.

tectiv3 commented:
I get that and agree, but then we are back to my original problem: 

history does not work with Content being a pointer...

meinside commented:
Oh, sorry for that.
Can I get your code? Or at least some snippet which can reproduce your problem.
I'd like to help you with that.

tectiv3 commented:
https://github.com/meinside/openai-go/pull/3#issuecomment-1596342592
This is it:)

It is from my https://github.com/tectiv3/chatgpt-bot repo.

tectiv3 commented:
Sorry for the confusion 
```go
type ChatMessage struct {
	ID        uint `gorm:"primarykey"`
	CreatedAt time.Time
	UpdatedAt time.Time
	ChatID    int64
	Role      openai.ChatMessageRole `json:"role"`
        Content *string `json:"content,omitempty"`
}

history := []openai.ChatMessage{system}
	for _, h := range chat.History {
		history = append(history, openai.ChatMessage{Role: h.Role, Content: h.Content})
	}

```
This is how it was using your version of the lib.

P.S.
I got an idea that maybe the problem is not in this library but in GORM not hydrating the value for Content when it is defined as a pointer.

meinside commented:
Hmm that's also strange. GORM supports pointer type values, but if it is not working since the change, let's see what's making the difference. Have you checked the saved values in the database, or the result of queries?

tectiv3 commented:
What I did:
- updated to 0.2.0
- noticed a change in model ChatMessage model
- updated my inherited model to use `*`
- updated history saving to pass by reference instead of value
- code compiled and I started playing with functions
- then I noticed that model does not remember the context when I talk to it
- tracked down the change to https://github.com/meinside/openai-go/blob/26a5357fe0453aa7adf9b6f234038e93ec810ec6/chat.go#L246 having empty content (the data was in present in DB)
- reverted the change in Content type
- everything went back to normal and I opened this PR

meinside commented:
Okay, let me write a sample program like your usage (load saved chat messages from GORM and send them all at once).
Please wait for a moment.

meinside commented:
```go
package main

import (
	"log"

	"github.com/meinside/openai-go"
	"gorm.io/driver/sqlite"
	"gorm.io/gorm"
)

const (
	dbPath = "db.sqlite"

	chatID = 111 // XXX: for testing

	modelName = "gpt-3.5-turbo-0613"

	systemMessage = "You are a helpful assistant."

	// XXX: insert yours here
	org = "org-xxxxxxxxxxx"
	key = "sk-yyyyyyyyyyyy"
)

type ChatMessage struct {
	gorm.Model
	ChatID  int64
	Role    openai.ChatMessageRole
	Content *string
}

func main() {
	if db, err := openDatabase(dbPath); err == nil {
		// XXX: inserting test data
		saveChatMessage(db, chatID, openai.NewChatUserMessage("Who won the world series in 2020?"))
		saveChatMessage(db, chatID, openai.NewChatAssistantMessage("The Los Angeles Dodgers won the World Series in 2020."))
		saveChatMessage(db, chatID, openai.NewChatUserMessage("Where was it played?"))

		if api := openai.NewClient(key, org); err == nil {
			system := openai.NewChatSystemMessage(systemMessage)

			history := []openai.ChatMessage{system}
			for _, m := range loadChatMessages(db, chatID) {
				history = append(history, openai.ChatMessage{Role: m.Role, Content: m.Content})
			}

			if res, err := api.CreateChatCompletion(
				modelName,
				history,
				openai.ChatCompletionOptions{}.
					SetFunctions([]openai.ChatCompletionFunction{
						openai.NewChatCompletionFunction(
							"gamePlayedAt",
							"A function to get the location where a game was played.",
							map[string]any{
								"type": "object",
								"properties": map[string]any{
									"location": map[string]any{
										"type":        "string",
										"description": "The name of the location.",
									},
									"latitude": map[string]any{
										"type":        "string",
										"description": "The latitude value of the location in float.",
									},
									"longitude": map[string]any{
										"type":        "string",
										"description": "The longitude value of the location in float.",
									},
								},
								"required": []string{"location", "latitude", "longitude"},
							},
						),
					}).
					SetFunctionCall(openai.ChatCompletionFunctionCallAuto)); err == nil {
				if len(res.Choices) > 0 {
					message := res.Choices[0].Message
					if message.FunctionCall != nil {
						functionName := message.FunctionCall.Name

						if functionName == "gamePlayedAt" {
							if args, err := message.FunctionCall.ArgumentsParsed(); err == nil {
								log.Printf("returned parameters for function: '%s' = %+v", functionName, args)
							} else {
								log.Printf("failed to parse arguments JSON: %s", err)
							}
						}
					}
				}
			} else {
				log.Printf("chat completion failed: %s", err)
			}
		}
	} else {
		log.Printf("failed to open database: %s", err)
	}
}

func openDatabase(dbPath string) (db *gorm.DB, err error) {
	db, err = gorm.Open(sqlite.Open(dbPath), &gorm.Config{
		PrepareStmt: true,
	})

	if err == nil {
		// migrate tables
		if err := db.AutoMigrate(
			&ChatMessage{},
		); err != nil {
			log.Printf("failed to migrate databases: %s", err)
		}

		// XXX: delete all rows before test
		db.Where("1 = 1").Delete(&ChatMessage{})

		return db, nil
	}

	return nil, err
}

func saveChatMessage(db *gorm.DB, chatID int64, chatMessage openai.ChatMessage) {
	if tx := db.Create(&ChatMessage{ChatID: chatID, Role: chatMessage.Role, Content: chatMessage.Content}); tx.Error != nil {
		log.Printf("failed to save chat message: %s", tx.Error)
	}
}

func loadChatMessages(db *gorm.DB, chatID int64) (result []ChatMessage) {
	if tx := db.Where("chat_id = ?", chatID).Order("id").Find(&result); tx.Error == nil {
		return result
	} else {
		log.Printf("failed to select chat messages: %s", tx.Error)
	}

	return []ChatMessage{}
}
```

I wrote a simple application that fetches chat messages from SQLite and send them to the API.

I did my best to make it look like yours, but there are trivial changes:

- removed JSON tags from the model
- added function call parameters for an example
- added some hard-coded values or function calls for testing (commented with `XXX`)

I could see it running as expected (inserting into & selecting from DB without any error),
but as I cannot see your whole code, I'm not sure what might be the problem on your side.

Hope it would help.

tectiv3 commented:
Well, technically you can :) https://github.com/tectiv3/chatgpt-bot

now it works... I don't know what it was... Sorry for wasting your time!
I literally just reverted to your lib, changed h.Content to *h.Content and it works :/

meinside commented:
Good to hear that! It was never a wasting of time :-)
